### PR TITLE
Bind `trace_ids` to the authorized experiment on `issues/invoke` and `genai/evaluate/invoke`

### PR DESCRIPTION
**Pull request by tanghaoji**
### Related Issues/PRs

Relates to GHSA-v7w2-x9m4-3743 (private security advisory). Follow-up to #25721, which fixed the same gap for `scorer/invoke`.

### What changes are proposed in this pull request?

Fixes a broken-object-authorization gap on `POST /ajax-api/3.0/mlflow/issues/invoke` and `POST /ajax-api/3.0/mlflow/genai/evaluate/invoke`.

Both routes are gated by `validate_can_update_experiment`, which only checks the caller's permission on the request's own `experiment_id`. Nothing bound the caller-supplied `trace_ids` to that experiment, so a user could name an experiment they control (they auto-get `MANAGE` on experiments they create) and supply trace ids from an experiment they cannot read. The submitted job would then fetch and analyze those traces and write the results into the caller's run. #25721 closed exactly this for `_invoke_scorer_handler` but did not touch the two sibling handlers.

Changes in `mlflow/server/handlers.py`:

- Extract the trace-binding check from `_invoke_scorer_handler` into a shared `_validate_trace_ids_in_experiment` helper. It batch-fetches the requested trace infos (with the per-trace `get_trace_info` fallback for stores that do not implement `batch_get_trace_infos`) and raises a generic, fail-closed `PERMISSION_DENIED` (403) if any existing trace resolves to a different experiment than the authorized one. Missing traces are not rejected here; they are left to fail downstream in the job, preserving the existing contract.
- Call the helper in `_invoke_issue_detection_handler` before credentials are fetched and before the run is created, and in `_invoke_genai_evaluate_handler` before scorer resolution and run creation. A rejected request therefore produces no run and no job.
- `_invoke_scorer_handler` now uses the helper, and its existing regression tests still pass.
- The three `/invoke` handlers are aligned on `trace_ids` handling: all reject an empty list and non-string items with a 400 (`issues/invoke` previously submitted a no-op run and job for `[]`; `scorer/invoke` previously returned a 500 for a non-string item), and all de-duplicate ids before the lookup. The fallback also triggers on a bare `NotImplementedError` from a plugin store; a store that implements neither lookup fails closed.

The guard lives in the handlers rather than the auth validators so it also protects deployments that mount the handlers without the basic-auth app, matching where the scorer guard already lives.

Reviewer note on impact: on a plain OSS basic-auth deployment the worker's callbacks (`link_traces_to_run`, `batch_get_traces`) go back through the auth layer and are rejected, so the demonstrated impact there is that a foreign trace id is accepted and a job is queued. Full content disclosure requires a deployment whose workers carry a service identity. The guard is cheap and keeps the three invoke endpoints consistent.

### How is this PR tested?

- [x] Existing unit/integration tests
- [x] New unit/integration tests
- [ ] Manual tests

New tests:

- `tests/server/test_handlers.py`: for each of `issues/invoke` and `genai/evaluate/invoke`, a cross-experiment rejection test (batch path), a fallback-path rejection test parametrized over `MlflowNotImplementedException` and bare `NotImplementedError`, a positive test where every trace is in the authorized experiment and the job is submitted, a de-duplication test, and a non-string `trace_ids` test. Plus an empty-`trace_ids` test for `issues/invoke`, a non-string `trace_ids` test for `scorer/invoke`, and direct unit tests of `_validate_trace_ids_in_experiment` covering same-experiment, foreign, missing, fallback, error-propagation, and fail-closed cases. The rejection tests assert that credential fetch, run creation, and job submission never happen. The rejection tests fail against the unpatched handlers, and the positive tests fail if the ownership check is inverted.
- `tests/server/jobs/test_genai_evaluate_invocation.py`: a real-server, real-store case where a trace from another experiment is rejected with a 403 and no run is created.
- `tests/server/auth/test_auth.py`: a basic-auth case (default-deny config) where a user who cannot read the victim's trace supplies it to each endpoint with their own experiment and gets a 403 from the handler, with no run created.

Pre-existing tests for these handlers gained the existing `mock_tracking_store` fixture because the handlers now consult the store before submitting (the same adjustment the scorer tests needed in #25721).

Verified locally:

- `tests/server/test_handlers.py`: 425 passed, 1 skipped
- `tests/server/auth/test_auth.py` (invoke endpoints): 7 passed
- `tests/server/jobs/test_genai_evaluate_invocation.py` (end-to-end, real server and SQL store): 6 passed

### Does this PR require documentation update?

- [x] No.

### Does this PR require updating the [MLflow Skills](https://github.com/mlflow/skills) repository?

- [x] No.

#### Is this a user-facing change?

- [x] No.

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [x] `area/tracking`: Tracking Service, tracking client APIs, autologging
- [x] `area/evaluation`: MLflow model evaluation features, evaluation metrics, and evaluation workflows

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [x] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes

#### Is this PR a critical bugfix or security fix that should go into the next patch release?

<details>
<summary>What is a minor/patch release?</summary>

- Minor release: a release that increments the second part of the version number (e.g., 1.2.0 -> 1.3.0).
  Minor releases are expected to contain larger changes, such as new features and improvements. Non-critical bug fixes and doc updates can be included as well. By default, your PR should target the next minor release.
- Patch release: a release that increments the third part of the version number (e.g., 1.2.0 -> 1.2.1).
  Patch releases are typically only performed when there has been a major regression or bug in the latest release. For the sake of stability, your PR should not be included in a patch release unless it is a critical fix, or if the risk level of your PR is exceedingly low.

</details>

- [ ] This PR is critical and needs to be in the next patch release
- [x] This PR can wait for the next minor release

This pull request and its description were written by Isaac.

### CODE REVIEW

File: mlflow/server/handlers.py [hunk 1]
```diff
@@ -5183,6 +5183,37 @@ def _set_review_queue_item_status():
...
+) -> None:
+    """
+    Reject the request if any requested trace that exists belongs to an experiment other
+    than ``experiment_id``. The route validators only check the caller's permission on the
+    request's ``experiment_id``, so this binds the caller-supplied ``trace_ids`` to that
+    authorized experiment. Missing traces are not rejected here; they are left to fail
+    downstream in the job, which preserves the existing contract for missing traces.
+    """
+    try:
+        trace_infos = tracking_store.batch_get_trace_infos(trace_ids)
```
[the remark this answers is not in the document]
Comment (tanghaoji) [reply mlflow/server/handlers.py:5296]: Added `_assert_item_type_string` to the `trace_ids` schema on both endpoints (matching the other trace-id handlers) so a non-string item now fails as a 400 validation error before the store is consulted, with a regression test for each endpoint (f9bac03423).

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

**Comment by tanghaoji** [comment]
/review

---

🚀 [NaiLaOpus running...](https://github.com/mlflow/mlflow/actions/runs/35039458277?pr=25873)

File: mlflow/server/handlers.py [hunk 2]
```diff
@@ -5331,6 +5364,7 @@ def _invoke_genai_evaluate_handler():
         )
```
[the remark this answers is not in the document]
Comment (tanghaoji) [reply mlflow/server/handlers.py:5472]: Widened the outer guard to catch bare `NotImplementedError` alongside `MlflowNotImplementedException` so a plugin store raising the built-in still takes the per-trace fallback, with a regression test (057e59b26b). The in-tree base class already raises `MlflowNotImplementedException` from `batch_get_trace_infos`, and every in-tree store lacking it implements `get_trace_info`. A store implementing neither lookup cannot prove ownership, so its error is deliberately left to propagate (fail closed) rather than letting the request through. Aligning the abstract base class is out of scope for this security fix.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: mlflow/server/handlers.py [hunk 3]
```diff
@@ -5245,9 +5278,11 @@ def _invoke_issue_detection_handler():
                 f"AI Gateway, or set {env_var_hint} on the MLflow server."
             )
 
+    store = _get_tracking_store()
+    _validate_trace_ids_in_experiment(store, trace_ids, experiment_id)
```
Comment (kriscon-db) [inline mlflow/server/handlers.py:5387]: An empty `trace_ids` reaches here and passes the guard vacuously (`_assert_required` permits `[]`), so a no-op run and job are still submitted. The genai handler (around 5369) and scorer (7254) both reject this earlier with `if not trace_ids: raise ...`. Worth adding the same guard here so the three endpoints stay consistent.

-- Claude-drafted, reviewed before posting.

**Review by kriscon-db — COMMENTED** [review]
Solid, well-scoped fix. Completeness checks out: the shared `_validate_trace_ids_in_experiment` helper now covers all three `/invoke` endpoints (scorer via #25721, plus these two), the guard runs before run creation and job submission on both handlers, it fails closed (foreign traces that exist surface with their real `experiment_id` and are rejected all-or-nothing, and the generic 403 does not leak ownership), and the other `trace_ids` endpoints bind ownership at the auth-validator layer, so the invoke family is complete. Nice extraction of the helper.

Two test-hardening asks before merge:
- The success/happy-path unit tests gained `mock_tracking_store` but never set `batch_get_trace_infos.return_value`, so the guard iterates an empty mock and passes vacuously. No test proves an all-same-experiment `trace_ids` set actually survives the guard and still submits, so a regression that inverted the check would keep every rejection test green. Could you add a positive case (traces all in `experiment_id`, assert `submit_job` is called) for the issues and genai handlers?
- The rejection tests all mock the store. A negative case in the real-server e2e (`tests/server/jobs/test_genai_evaluate_invocation.py`: a READ-denied user supplies a foreign trace_id, expect 403, no run/job) would lock the auth boundary in at the integration layer where the real validator and handler agree on the same `experiment_id`.

Consistency across the three endpoints this PR is aligning (one inline):
- The scorer handler still reads `trace_ids` without `_assert_item_type_string`, so a non-string id there 500s instead of a clean 400, while the two new handlers are hardened.
- Minor: scorer dedups `trace_ids` before the check; the two new call sites do not (perf/consistency only).

Optional: the two fallback-exception tests (`MlflowNotImplementedException` vs bare `NotImplementedError`) could fold into one `@pytest.mark.parametrize`, and a direct unit test of `_validate_trace_ids_in_experiment` over the matrix would be cleaner than exercising it only through three endpoints.

-- Claude-drafted, reviewed before posting.

File: mlflow/server/handlers.py (hunk 3 above)
Comment (tanghaoji) [reply mlflow/server/handlers.py:5387]: Added the same `if not trace_ids` guard to `_invoke_issue_detection_handler`, so an empty list is a 400 before any run or job is created (the UI never submits an empty selection), with `test_invoke_issue_detection_handler_rejects_empty_trace_ids` (0d461a9d6). The same commit covers the review-summary asks: positive same-experiment tests for both handlers (verified to fail if the check is inverted), a real-server negative case in the genai evaluate e2e suite plus a basic-auth READ-denied-user case for both endpoints in `tests/server/auth/test_auth.py`, string-item validation on `scorer/invoke`, de-duplication at the two new call sites, the parametrized fallback tests, and direct unit tests of `_validate_trace_ids_in_experiment`.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

**Review by kriscon-db — APPROVED** [review]
Re-reviewed the follow-ups (0d461a9d, 6990bca0): everything's addressed. The three `/invoke` handlers are aligned (issue-detection now guards empty `trace_ids`; all three dedup; the scorer rejects non-string `trace_ids` with a clean 400), the dedup runs before `_validate_trace_ids_in_experiment` so the security binding is unchanged, and the tests now cover the positive same-experiment path (`submit_job` asserted called), a real-server negative case for a foreign-experiment trace, and direct unit tests of the shared helper. No new concerns. Approving.

-- Claude-drafted, reviewed before posting.
